src/types/collectors.rs

```diff
@@ -0,0 +1,155 @@
+/*
+    Copyright © 2023, ParallelChain Lab
+    Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+//! Defines the generic [`SignedMessage`] and [`Collector`] traits. 
+//! 
+//! Implementations used by the [Pacemaker][crate::pacemaker::types] and 
+//! [HotStuff][crate::hotstuff::types] protocols can be found in the respective modules. [`Collectors`]
+//! groups collectors for all active validator sets into a single struct, which can be easily updated
+//! on view or validator set state updates.
+
+pub use ed25519_dalek::{SigningKey, VerifyingKey, Signature};
+pub use sha2::Sha256 as CryptoHasher;
+
+use crate::messages::SignedMessage;
+use crate::state::{
+    block_tree::{BlockTree, BlockTreeError}, 
+    kv_store::KVStore
+};
+
+use super::{
+    basic::{ChainID, TotalPower, ViewNumber}, 
+    validators::{ValidatorSet, ValidatorSetState}
+};
+
+/// Evidence that a quorum of validators from a given validator set supports a given decision. The 
+/// evidence comes in the form of a set of signatures of the validators.
+pub trait Certificate {
+
+    /// Returns whether the certificate is correct, i.e., can serve as evidence that a given block is 
+    /// approved by a quorum of validators assigned to validate it. This may require checking if it is 
+    /// correctly signed by a quorum of validators from an appropriate validator set.
+    fn is_correct<K: KVStore>(&self, block_tree: &BlockTree<K>) -> Result<bool, BlockTreeError>;
+
+    /// Returns whether the certificate is correctly signed by a quorum of validators from the given 
+    /// validator set.
+    fn is_correctly_signed(&self, validator_set: &ValidatorSet) -> bool;
```

Comment:
This function seems to only be used within the implementation of `is_correct` and I can't really see other code using this function directly. 

Is it then worth it to keep this as a dedicated method in trait? Doesn't it kind of confuse the public interface of the trait?

Reply:
Indeed it's meant to be called by `is_correct`, not used directly. But I don't think it's confusing, well assuming I explained in the comments what the purpose of having both is, perhaps I didn't. `is_correctly_signed` is the old `is_correct` which checks cryptographic correctness against a given validator set - it is important to implement this validation logic for any certificate. `is_correct` also figures out which validator set should have signed a given qc, and then calls `is_correctly_signed` to validate the qc against this validator set.